File: packages/core/test/samples/sample.ts
import { readFileSync } from "node:fs";
import type { Config } from "./config.js";

export interface Entry {
  date: Date;
  cents: number;
  category: string;
  note?: string;
}

export type Section = "essentials" | "lifestyle" | "other";

export const ALIASES: Record<string, string> = {
  groceries: "food",
  bus: "transport",
};

export class Ledger {
  private entries: Entry[] = [];
  private keys = new Set<string>();

  add(entry: Entry): boolean {
    const key = `${entry.date.toISOString()}|${entry.cents}|${entry.category}|${entry.note ?? ""}`;
    if (this.keys.has(key)) return false;
    this.keys.add(key);
    this.entries.push(entry);
    return true;
  }

  total(category?: string): number {
    return this.entries.filter((e) => !category || e.category === category).reduce((a, e) => a + e.cents, 0);
  }
}

export function formatMoney(cents: number, config?: Config): string {
  if (config) return new Intl.NumberFormat(config.locale, { style: "currency", currency: config.currency }).format(cents / 100);
  return (cents / 100).toFixed(2);
}

export default function loadConfig(path: string): Config {
  try {
    return JSON.parse(readFileSync(path, "utf8"));
  } catch {
    return { currency: "USD", locale: "en-US" };
  }
}

File: packages/core/test/samples/sample.rs
use std::collections::HashSet;

/// An entry in the ledger.
#[derive(Debug, Clone, PartialEq)]
pub struct Entry {
    pub date: String,
    pub cents: i64,
    pub category: String,
    pub note: String,
}

pub enum Section {
    Essentials,
    Lifestyle,
    Other,
}

pub trait Format {
    fn format_money(&self, cents: i64) -> String;
}

#[derive(Default)]
pub struct Ledger {
    entries: Vec<Entry>,
    keys: HashSet<String>,
}

impl Ledger {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add(&mut self, entry: Entry) -> bool {
        let key = format!("{}|{}|{}|{}", entry.date, entry.cents, entry.category, entry.note);
        if !self.keys.insert(key) {
            return false;
        }
        self.entries.push(entry);
        true
    }

    pub fn total(&self, category: Option<&str>) -> i64 {
        self.entries
            .iter()
            .filter(|e| category.map_or(true, |c| e.category == c))
            .map(|e| e.cents)
            .sum()
    }
}

pub fn format_money(cents: i64) -> String {
    let sign = if cents < 0 { "-" } else { "" };
    let a = cents.abs();
    format!("{}{}.{:02}", sign, a / 100, a % 100)
}

const MAX_ROWS: usize = 1000;
